### PR TITLE
gradientDifference could be an array and damping parameter is automatically updated

### PR DESCRIPTION
**Pull request by jobo322**
gradientDifference could have a value for each parameter, so the space could be explored better. Also if the gradientDifference value is equal to zero the parameter will not be updated.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -63,6 +66,17 @@ export default function levenbergMarquardt(
...
+    isArray(gradientDifference) &&
+    gradientDifference.length !== parameters.length
+  ) {
+    gradientDifference = new Array(parameters.length).fill(
+      gradientDifference[0],
+    );
+  } else if (typeof gradientDifference === 'number') {
+    gradientDifference = new Array(parameters.length).fill(gradientDifference);
+  }
+
```
Comment (lpatiny) [inline src/index.js:79]: Please throw if none of those cases

File: src/index.js [hunk 2]
```diff
@@ -8,8 +8,10 @@ import step from './step';
  * @param {{x:Array<number>, y:Array<number>}} data - Array of points to fit in the format [x1, x2, ... ], [y1, y2, ... ]
  * @param {function} parameterizedFunction - The parameters and returns a function with the independent variable as a parameter
  * @param {object} [options] - Options object
- * @param {number} [options.damping] - Levenberg-Marquardt parameter
- * @param {number} [options.gradientDifference = 10e-2] - Adjustment for decrease the damping parameter
+ * @param {number} [options.damping = 0] - Levenberg-Marquardt parameter, small values of the damping parameter λ result in a Gauss-Newton update and large
+values of λ result in a gradient descent update
+ * @param {number|array} [options.gradientDifference = 10e-2] - The step size to approximate the jacobian matrix
+ * @param {boolean} [options.centralDifference = false] - If true the jacobian matrix is approximated by central differences otherwise by forward differences
```
Comment (lpatiny) [inline src/index.js:14]: Please add in the README a good explanation about this difference.

**Comment by lpatiny** [comment]
Thanks for those improvements. Did yout ry this code in a real case ?

Could you try to improve the README with an example of the new options ?

Could you confirm that there is no breaking changes and the code yields exactly the same results at the same speed as previously ?

File: src/step.js [hunk 3]
```diff
@@ -87,18 +55,27 @@ export default function step(
...
   );
   let matrixFunc = matrixFunction(data, evaluatedData);
+  // console.log(' identityt' , identity)
+  // console.log(' gradient 2', identity.add(gradientFunc.mmul(gradientFunc.transpose())));
   let inverseMatrix = inverse(
     identity.add(gradientFunc.mmul(gradientFunc.transpose())),
   );
 
   params = new Matrix([params]);
+  // console.log(
```
Comment (lpatiny) [inline src/step.js:68]: Please remove comments

File: src/step.js [hunk 4]
```diff
@@ -87,18 +55,27 @@ export default function step(
     params,
     gradientDifference,
     parameterizedFunction,
+    centralDifference,
   );
   let matrixFunc = matrixFunction(data, evaluatedData);
+  // console.log(' identityt' , identity)
```
Comment (lpatiny) [inline src/step.js:61]: Please remove those comments

File: src/index.js (hunk 2 above)
Comment (lpatiny) [reply src/index.js:14]: Do you have a reference / link to literature where this kind of approach is used ?

File: src/step.js [hunk 5]
```diff
@@ -87,18 +55,27 @@ export default function step(
...
+  //   .mmul(gradientFunc)
+  //   .mmul(matrixFunc)
+  //   .transpose()
+  // );
+
   params = params.sub(
     inverseMatrix
       .mmul(gradientFunc)
       .mmul(matrixFunc)
-      .mul(gradientDifference)
```
Comment (lpatiny) [inline src/step.js:101]: I don't know how this algorithm exactly works but it is surprising we can remove this line.

File: src/index.js (hunk 1 above)
Comment (jobo322) [reply src/index.js:44]: it is done

File: src/index.js (hunk 2 above)
Comment (jobo322) [reply src/index.js:17]: yes, the condensed is one is [http://people.duke.edu/~hpgavin/ce281/lm.pdf](http://people.duke.edu/~hpgavin/ce281/lm.pdf)

File: src/step.js (hunk 3 above)
Comment (jobo322) [reply src/step.js]: it is done

File: src/step.js (hunk 5 above)
Comment (jobo322) [reply src/step.js:101]: it is because this gradientDifference value now is inside of gradientFunc due it could have a value for each parameter

**Comment by jobo322** [comment]
Now the damping parameters is updated on the fly, it could accelerate the convergence and it was testing with the aromatic part of 1D-1H-NMR spectrum.

File: src/index.js (hunk 2 above)
Comment (lpatiny) [reply src/index.js:17]: You should add this reference in the README as well

**Pull request merged**